pkg/volume/flexvolume/flexvolume.go

```diff
@@ -241,76 +234,142 @@ func (f flexVolumeMounter) GetAttributes() volume.Attributes {
+	unmount(unmounter *flexVolumeUnmounter, dir string) error
 }
 
 // SetUpAt creates new directory.
 func (f *flexVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
-
-	notmnt, err := f.blockDeviceMounter.IsLikelyNotMountPoint(dir)
-	if err != nil && !os.IsNotExist(err) {
-		glog.Errorf("Cannot validate mountpoint: %s", dir)
+	err := f.Attach(f.spec, "")
+	if err != nil {
+		return err
+	}
+	devicePath, err := f.WaitForAttach(f.spec, 1*time.Hour)
+	if err != nil {
 		return err
 	}
-	if !notmnt {
-		return nil
+	deviceMountPath := f.GetDeviceMountPath(f.spec)
+	return f.MountDevice(f.spec, devicePath, deviceMountPath, f.mounter)
+}
+
+type attachResult struct {
+	device string
+	err    error
+}
+
+func getVolumeSource(spec *volume.Spec) *api.FlexVolumeSource {
+	if spec.Volume != nil {
+		return spec.Volume.FlexVolume
 	}
+	return spec.PersistentVolume.Spec.FlexVolume
+}
 
-	if f.options == nil {
-		f.options = make(map[string]string)
+func (f *flexVolumeMounter) getVolumeOptions(spec *volume.Spec) (map[string]string, error) {
+	volumeSource := getVolumeSource(spec)
+	options := make(map[string]string)
+	for name, value := range volumeSource.Options {
+		options[name] = value
 	}
 
-	f.options[optionFSType] = f.fsType
+	options[optionFSType] = volumeSource.FSType
 
 	// Read write mount options.
-	if f.readOnly {
-		f.options[optionReadWrite] = "ro"
+	if spec.ReadOnly {
+		options[optionReadWrite] = "ro"
 	} else {
-		f.options[optionReadWrite] = "rw"
+		options[optionReadWrite] = "rw"
 	}
 
 	// Extract secret and pass it as options.
-	for name, secret := range f.secrets {
-		f.options[optionKeySecret+"/"+name] = secret
+	secrets, err := getSecrets(spec, spec.PodNamespace, f.plugin.host)
+	if err != nil {
+		return nil, err
+	}
+	for name, secret := range secrets {
+		options[optionKeySecret+"/"+name] = secret
 	}
+	return options, nil
+}
 
-	device, err := f.manager.attach(f)
+func (f *flexVolumeMounter) Attach(spec *volume.Spec, hostName string) error {
+	options, err := f.getVolumeOptions(spec)
 	if err != nil {
-		if !isCmdNotSupportedErr(err) {
-			glog.Errorf("Failed to attach volume: %s", f.volName)
-			return err
+		return err
+	}
+
+	if f.attachResultC == nil {
+		f.attachResultC = make(chan attachResult, 1)
+	}
+	go func() {
+		device, err := f.manager.attach(f, options, hostName)
+		f.attachResultC <- attachResult{
+			device: device,
+			err:    err,
+		}
+	}()
+	return nil
+}
+
+func (f *flexVolumeMounter) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
+	if f.attachResultC == nil {
+		// attach wasn't called
+		return "", fmt.Errorf("Attach wasn't called")
+	}
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	select {
+	case r := <-f.attachResultC:
+		if r.err != nil && !isCmdNotSupportedErr(r.err) {
+			glog.Errorf("Failed to attach volume: %s", spec.Name())
+			return "", r.err
 		}
 		// Attach not supported or required. Continue to mount.
+		return r.device, nil
+	case <-timer.C:
```

Comment:
Kill the pending "attach" call/invocation on timeout. That should take care of the process leak on timeout.